stdlib/public/core/Collection.swift

```diff
@@ -56,8 +57,67 @@ public protocol Indexable {
   ///
   /// - Complexity: O(1)
   subscript(position: Index) -> _Element { get }
+  
+// TODO: swift-3-indexing-model - replaces functionality in ForwardIndex
+  /// Performs a range check in O(1), or a no-op when a range check is not
+  /// implementable in O(1).
+  ///
+  /// The range check, if performed, is equivalent to:
+  ///
+  ///     precondition(bounds.contains(index))
+  ///
+  /// Use this function to perform a cheap range check for QoI purposes when
+  /// memory safety is not a concern.  Do not rely on this range check for
+  /// memory safety.
+  ///
+  /// The default implementation for forward and bidirectional indices is a
+  /// no-op.  The default implementation for random access indices performs a
+  /// range check.
+  ///
+  /// - Complexity: O(1).
+  func _failEarlyRangeCheck(index: Index, bounds: Range<Index>)
+  
+// TODO: swift-3-indexing-model - replaces functionality in ForwardIndex
+  /// Performs a range check in O(1), or a no-op when a range check is not
+  /// implementable in O(1).
+  ///
+  /// The range check, if performed, is equivalent to:
+  ///
+  ///     precondition(
+  ///       bounds.contains(range.startIndex) ||
+  ///       range.startIndex == bounds.endIndex)
+  ///     precondition(
+  ///       bounds.contains(range.endIndex) ||
+  ///       range.endIndex == bounds.endIndex)
+  ///
+  /// Use this function to perform a cheap range check for QoI purposes when
+  /// memory safety is not a concern.  Do not rely on this range check for
+  /// memory safety.
+  ///
+  /// The default implementation for forward and bidirectional indices is a
+  /// no-op.  The default implementation for random access indices performs a
+  /// range check.
+  ///
+  /// - Complexity: O(1).
+  func _failEarlyRangeCheck(
+    rangeStart rangeStart: Index,
+    rangeEnd: Index,
+    boundsStart: Index,
+    boundsEnd: Index)
```

Comment:
I can try making that change in the other PR which will contain unittest breaking changes as things are cut over to Collection off of ForwardIndex.